File: sydney.py
import base64
import json
import os
import random
import urllib.request
import uuid
from enum import Enum
from time import time
from typing import Union

import aiohttp

_PROXY = urllib.request.getproxies().get("https")

_BASE_OPTION_SETS = [
    "nlu_direct_response_filter",
    "deepleo",
    "disable_emoji_spoken_text",
    "responsible_ai_policy_235",
    "enablemm",
    "iycapbing",
    "iyxapbing",
    "dv3sugg",
    "iyoloxap",
    "iyoloneutral",
    "gencontentv3",
    "nojbfedge",
]


class _OptionSets(Enum):
    CREATIVE = _BASE_OPTION_SETS + ["h3imaginative"]
    BALANCED = _BASE_OPTION_SETS + ["galileo"]
    PRECISE = _BASE_OPTION_SETS + ["h3precise"]


_SLICE_IDS = [
    "winmuid1tf",
    "newmma-prod",
    "imgchatgptv2",
    "tts2",
    "voicelang2",
    "anssupfotest",
    "emptyoson",
    "tempcacheread",
    "temptacache",
    "ctrlworkpay",
    "winlongmsg2tf",
    "628fabocs0",
    "531rai268s0",
    "602refusal",
    "621alllocs0",
    "621docxfmtho",
    "621preclsvn",
    "330uaug",
    "529rweas0",
    "0626snptrcs0",
    "619dagslnv1nr"
]


class _LocationHint(Enum):
    USA = {
        "locale": "en-US",
        "LocationHint": [
            {
                "country": "United States",
                "state": "California",
                "city": "Los Angeles",
                "timezoneoffset": 8,
                "countryConfidence": 8,
                "Center": {
                    "Latitude": 34.0536909,
                    "Longitude": -118.242766,
                },
                "RegionType": 2,
                "SourceType": 1,
            },
        ],
    }
    CHINA = {
        "locale": "zh-CN",
        "LocationHint": [
            {
                "country": "China",
                "state": "",
                "city": "Beijing",
                "timezoneoffset": 8,
                "countryConfidence": 8,
                "Center": {
                    "Latitude": 39.9042,
                    "Longitude": 116.4074,
                },
                "RegionType": 2,
                "SourceType": 1,
            },
        ],
    }
    EU = {
        "locale": "en-IE",
        "LocationHint": [
            {
                "country": "Norway",
                "state": "",
                "city": "Oslo",
                "timezoneoffset": 1,
                "countryConfidence": 8,
                "Center": {
                    "Latitude": 59.9139,
                    "Longitude": 10.7522,
                },
                "RegionType": 2,
                "SourceType": 1,
            },
        ],
    }
    UK = {
        "locale": "en-GB",
        "LocationHint": [
            {
                "country": "United Kingdom",
                "state": "",
                "city": "London",
                "timezoneoffset": 0,
                "countryConfidence": 8,
                "Center": {
                    "Latitude": 51.5074,
                    "Longitude": -0.1278,
                },
                "RegionType": 2,
                "SourceType": 1,
            },
        ],
    }


_DELIMITER = '\x1e'
_FORWARDED_IP = f"1.0.0.{random.randint(0, 255)}"

_ALLOWED_MESSAGE_TYPES = [
    "ActionRequest",
    "Chat",
    "Context",
    "InternalSearchQuery",
    "InternalSearchResult",
    "Disengaged",
    "InternalLoaderMessage",
    "Progress",
    "RenderCardRequest",
    "AdsQuery",
    "SemanticSerp",
    "GenerateContentQuery",
    "SearchQuery"
]

_HEADERS = {
    "accept": "application/json",
    "accept-language": "en-US,en;q=0.9",
    "content-type": "application/json",
    "sec-ch-ua": '"Not_A Brand";v="99", Microsoft Edge";v="110", "Chromium";v="110"',
    "sec-ch-ua-arch": '"x86"',
    "sec-ch-ua-bitness": '"64"',
    "sec-ch-ua-full-version": '"109.0.1518.78"',
    "sec-ch-ua-full-version-list": '"Chromium";v="110.0.5481.192", "Not A(Brand";v="24.0.0.0", "Microsoft Edge";v="110.0.1587.69"',
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-model": "",
    "sec-ch-ua-platform": '"Windows"',
    "sec-ch-ua-platform-version": '"15.0.0"',
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "same-origin",
    "x-ms-client-request-id": str(uuid.uuid4()),
    "x-ms-useragent": "azsdk-js-api-client-factory/1.0.0-beta.1 core-rest-pipeline/1.10.0 OS/Win32",
    "Referer": "https://www.bing.com/search?q=Bing+AI&showconv=1&FORM=hpcodx",
    "Referrer-Policy": "origin-when-cross-origin",
    "x-forwarded-for": _FORWARDED_IP,
}

_HEADERS_INIT_CONVER = {
    "authority": "www.bing.com",
    "accept": "application/json",
    "accept-language": "en-US,en;q=0.9",
    "cache-control": "max-age=0",
    "sec-ch-ua": '"Chromium";v="110", "Not A(Brand";v="24", "Microsoft Edge";v="110"',
    "sec-ch-ua-arch": '"x86"',
    "sec-ch-ua-bitness": '"64"',
    "sec-ch-ua-full-version": '"110.0.1587.69"',
    "sec-ch-ua-full-version-list": '"Chromium";v="110.0.5481.192", "Not A(Brand";v="24.0.0.0", "Microsoft Edge";v="110.0.1587.69"',
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-model": '""',
    "sec-ch-ua-platform": '"Windows"',
    "sec-ch-ua-platform-version": '"15.0.0"',
    "upgrade-insecure-requests": "1",
    "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36 Edg/112.0.1722.46",
    "x-edge-shopping-flag": "1",
    "x-forwarded-for": _FORWARDED_IP,
}


def _format(msg: dict) -> str:
    return json.dumps(msg, ensure_ascii=False) + _DELIMITER


async def create_conversation(
        proxy: str | None = _PROXY,
        cookies: list[dict] | None = None,
):
    formatted_cookies = {}
    if cookies:
        for cookie in cookies:
            formatted_cookies[cookie["name"]] = cookie["value"]
    async with aiohttp.ClientSession(
            cookies=formatted_cookies,
            headers=_HEADERS_INIT_CONVER,
    ) as session:
        response = await session.get(
            url="https://edgeservices.bing.com/edgesvc/turing/conversation/create",
            proxy=proxy,
        )
    if response.status != 200:
        text = await response.text()
        raise Exception(f"Authentication failed {text}")
    try:
        conversation = await response.json()
    except:
        text = await response.text()
        raise Exception(text)
    if conversation["result"]["value"] == "UnauthorizedRequest":
        raise Exception(conversation["result"]["message"])
    return conversation


def _get_location_hint_from_locale(locale: str) -> Union[dict, None]:
    locale = locale.lower()
    if locale == "en-gb":
        hint = _LocationHint.UK.value
    elif locale == "en-ie":
        hint = _LocationHint.EU.value
    elif locale == "zh-cn":
        hint = _LocationHint.CHINA.value
    else:
        hint = _LocationHint.USA.value
    return hint.get("LocationHint")


async def ask_stream(
        conversation: dict,
        prompt: str,
        context: str,
        conversation_style: str = "creative",
        locale: str = "zh-CN",
        proxy=_PROXY,
        image_url=None,
):
    timeout = aiohttp.ClientTimeout(total=900)
    async with aiohttp.ClientSession(timeout=timeout) as session:
        conversation_id = conversation["conversationId"]
        client_id = conversation["clientId"]
        conversation_signature = conversation["conversationSignature"]

        async with session.ws_connect(
                'wss://sydney.bing.com/sydney/ChatHub',
                autoping=False,
                headers=_HEADERS,
                proxy=proxy
        ) as wss:
            await wss.send_str(_format({'protocol': 'json', 'version': 1}))
            await wss.receive(timeout=900)
            await wss.send_str(_format({"type": 6}))

            struct = {
                'arguments': [
                    {
                        'optionsSets': getattr(_OptionSets, conversation_style.upper()).value,
                        'source': 'cib',
                        'allowedMessageTypes': _ALLOWED_MESSAGE_TYPES,
                        'sliceIds': _SLICE_IDS,
                        'traceId': os.urandom(16).hex(),
                        'isStartOfSession': True,
                        'message': {
                            "locale": locale,
                            "market": locale,
                            "region": locale[-2:],  # en-US -> US
                            "locationHints": _get_location_hint_from_locale(locale),
                            "author": "user",
                            "inputMethod": "Keyboard",
                            "text": prompt,
                            "messageType": random.choice(["Chat", "SearchQuery"]),
                            "imageUrl": image_url or None,
                        },
                        'conversationSignature': conversation_signature,
                        'participant': {
                            'id': client_id
                        },
                        'conversationId': conversation_id,
                        'previousMessages': [
                            {
                                "author": "user",
                                "description": context,
                                "contextType": "WebPage",
                                "messageType": "Context",
                            },
                        ]
                    }
                ],
                'invocationId': '0',
                'target': 'chat',
                'type': 4
            }

            await wss.send_str(_format(struct))

            retry_count = 5
            while True:
                if wss.closed:
                    break
                msg = await wss.receive(timeout=900)

                if not msg.data:
                    retry_count -= 1
                    if retry_count == 0:
                        raise Exception("No response from server")
                    continue

                if isinstance(msg.data, str):
                    objects = msg.data.split(_DELIMITER)
                else:
                    continue

                for obj in objects:
                    if int(time()) % 6 == 0:
                        await wss.send_str(_format({"type": 6}))
                    if not obj:
                        continue
                    response = json.loads(obj)
                    if response["type"] == 2:
                        if response["item"]["result"].get("error"):
                            raise Exception(
                                f"{response['item']['result']['value']}: {response['item']['result']['message']}")
                    yield response
                    if response["type"] == 2:
                        break


async def upload_image(filename, proxy):
    async with aiohttp.ClientSession(
            headers={"Referer": "https://www.bing.com/search?q=Bing+AI&showconv=1&FORM=hpcodx"}
    ) as session:
        url = "https://www.bing.com/images/kblob"

        payload = {
            "imageInfo": {},
            "knowledgeRequest": {
                "invokedSkills": ["ImageById"],
                "subscriptionId": "Bing.Chat.Multimodal",
                "invokedSkillsRequestData": {"enableFaceBlur": False},
                "convoData": {
                    "convoid": "",
                    "convotone": "Creative"
                }
            }
        }

        with open(filename, 'rb') as f:
            file_data = f.read()
            image_base64 = base64.b64encode(file_data)

        data = aiohttp.FormData()
        data.add_field('knowledgeRequest', json.dumps(payload), content_type="application/json")
        data.add_field('imageBase64', image_base64.decode('utf-8'), content_type="application/octet-stream")

        async with session.post(url, data=data, proxy=proxy) as resp:
            return (await resp.json())["blobId"]

File: sydney.go
package main

import (
	"context"
	"encoding/json"
	"github.com/google/uuid"
	"github.com/pkg/errors"
	"github.com/tidwall/gjson"
	"io"
	"log"
	"net/http"
	"net/url"
	"nhooyr.io/websocket"
	"strconv"
	"strings"
	"time"
)

const delimiter = '\x1e'

type Sydney struct {
	debug             bool
	cookies           map[string]string
	proxy             string
	conversationStyle string
	locale            string
	wssURL            string
	noSearch          bool

	basicOptionsSet           []string
	optionsSetMap             map[string][]string
	sliceIDs                  []string
	locationHints             map[string][]LocationHint
	allowedMessageTypes       []string
	headers                   map[string]string
	headersCreateConversation map[string]string
}
type LocationHint struct {
	Country           string `json:"country"`
	State             string `json:"state"`
	City              string `json:"city"`
	TimezoneOffset    int    `json:"timezoneoffset"`
	CountryConfidence int    `json:"countryConfidence"`
	Center            LatLng `json:"Center"`
	RegionType        int    `json:"RegionType"`
	SourceType        int    `json:"SourceType"`
}
type LatLng struct {
	Latitude  float64 `json:"Latitude"`
	Longitude float64 `json:"Longitude"`
}
type CreateConversationResult struct {
	Value   string `json:"value"`
	Message string `json:"message"`
}
type CreateConversationResponse struct {
	ConversationId        string                   `json:"conversationId"`
	ClientId              string                   `json:"clientId"`
	Result                CreateConversationResult `json:"result"`
	SecAccessToken        string                   `json:"secAccessToken"`
	ConversationSignature string                   `json:"conversationSignature"`
}

func NewSydney(debug bool, cookies map[string]string, proxy string,
	conversationStyle string, locale string, wssDomain string, noSearch bool) *Sydney {
	basicOptionsSet := []string{
		"nlu_direct_response_filter",
		"deepleo",
		"disable_emoji_spoken_text",
		"responsible_ai_policy_235",
		"enablemm",
		"iycapbing",
		"iyxapbing",
		"dv3sugg",
		"iyoloxap",
		"iyoloneutral",
		"gencontentv3",
		"nojbfedge",
	}
	uuidObj, err := uuid.NewUUID()
	if err != nil {
		panic(err)
	}
	forwardedIP := "1.0.0." + strconv.Itoa(RandIntInclusive(1, 255))
	return &Sydney{
		debug:             debug,
		cookies:           cookies,
		proxy:             proxy,
		conversationStyle: conversationStyle,
		locale:            locale,
		wssURL: Ternary(wssDomain == "", "wss://sydney.bing.com/sydney/ChatHub",
			"wss://"+wssDomain+"/sydney/ChatHub"),
		noSearch:        noSearch,
		basicOptionsSet: basicOptionsSet,
		optionsSetMap: map[string][]string{
			"Creative": append(basicOptionsSet, "h3imaginative"),
			"Balanced": append(basicOptionsSet, "galileo"),
			"Precise":  append(basicOptionsSet, "h3precise"),
		},
		sliceIDs: []string{
			"winmuid1tf",
			"newmma-prod",
			"imgchatgptv2",
			"tts2",
			"voicelang2",
			"anssupfotest",
			"emptyoson",
			"tempcacheread",
			"temptacache",
			"ctrlworkpay",
			"winlongmsg2tf",
			"628fabocs0",
			"531rai268s0",
			"602refusal",
			"621alllocs0",
			"621docxfmtho",
			"621preclsvn",
			"330uaug",
			"529rweas0",
			"0626snptrcs0",
			"619dagslnv1nr",
		},
		locationHints: map[string][]LocationHint{
			"zh-CN": {
				{
					Country:           "China",
					State:             "",
					City:              "Beijing",
					TimezoneOffset:    8,
					CountryConfidence: 8,
					Center: LatLng{
						Latitude:  39.9042,
						Longitude: 116.4074,
					},
					RegionType: 2,
					SourceType: 1,
				},
			},
			"en-US": {
				{
					Country:           "United States",
					State:             "California",
					City:              "Los Angeles",
					TimezoneOffset:    8,
					CountryConfidence: 8,
					Center: LatLng{
						Latitude:  34.0536909,
						Longitude: -118.242766,
					},
					RegionType: 2,
					SourceType: 1,
				},
			},
		},
		allowedMessageTypes: []string{
			"ActionRequest",
			"Chat",
			"Context",
			"InternalSearchQuery",
			"InternalSearchResult",
			"Disengaged",
			"InternalLoaderMessage",
			"Progress",
			"RenderCardRequest",
			"AdsQuery",
			"SemanticSerp",
			"GenerateContentQuery",
			"SearchQuery",
		},
		headers: map[string]string{
			"accept":                      "application/json",
			"accept-language":             "en-US,en;q=0.9",
			"content-type":                "application/json",
			"sec-ch-ua":                   "\"Not_A Brand\";v=\"99\", Microsoft Edge\";v=\"110\", \"Chromium\";v=\"110\"",
			"sec-ch-ua-arch":              "\"x86\"",
			"sec-ch-ua-bitness":           "\"64\"",
			"sec-ch-ua-full-version":      "\"109.0.1518.78\"",
			"sec-ch-ua-full-version-list": "\"Chromium\";v=\"110.0.5481.192\", \"Not A(Brand\";v=\"24.0.0.0\", \"Microsoft Edge\";v=\"110.0.1587.69\"",
			"sec-ch-ua-mobile":            "?0",
			"sec-ch-ua-model":             "",
			"sec-ch-ua-platform":          "\"Windows\"",
			"sec-ch-ua-platform-version":  "\"15.0.0\"",
			"sec-fetch-dest":              "empty",
			"sec-fetch-mode":              "cors",
			"sec-fetch-site":              "same-origin",
			"x-ms-client-request-id":      uuidObj.String(),
			"x-ms-useragent":              "azsdk-js-api-client-factory/1.0.0-beta.1 core-rest-pipeline/1.10.0 OS/Win32",
			"Referer":                     "https://www.bing.com/search?q=Bing+AI&showconv=1&FORM=hpcodx",
			"Referrer-Policy":             "origin-when-cross-origin",
			"x-forwarded-for":             forwardedIP,
		},
		headersCreateConversation: map[string]string{
			"authority":                   "www.bing.com",
			"accept":                      "application/json",
			"accept-language":             "en-US,en;q=0.9",
			"cache-control":               "max-age=0",
			"sec-ch-ua":                   `"Chromium";v="110", "Not A(Brand";v="24", "Microsoft Edge";v="110"`,
			"sec-ch-ua-arch":              `"x86"`,
			"sec-ch-ua-bitness":           `"64"`,
			"sec-ch-ua-full-version":      `"110.0.1587.69"`,
			"sec-ch-ua-full-version-list": `"Chromium";v="110.0.5481.192", "Not A(Brand";v="24.0.0.0", "Microsoft Edge";v="110.0.1587.69"`,
			"sec-ch-ua-mobile":            `"?0"`,
			"sec-ch-ua-model":             `""`,
			"sec-ch-ua-platform":          `"Windows"`,
			"sec-ch-ua-platform-version":  `"15.0.0"`,
			"upgrade-insecure-requests":   "1",
			"user-agent":                  "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36 Edg/112.0.1722.46",
			"x-edge-shopping-flag":        "1",
			"x-forwarded-for":             forwardedIP,
		},
	}
}

func (o *Sydney) CreateConversation() (CreateConversationResponse, error) {
	client, err := MakeHTTPClient(o.proxy, 10*time.Second)
	if err != nil {
		return CreateConversationResponse{}, err
	}
	req, err := http.NewRequest("GET", "https://edgeservices.bing.com/edgesvc/turing/conversation/create", nil)
	if err != nil {
		return CreateConversationResponse{}, err
	}
	req.Header.Set("Cookie", FormatCookieString(o.cookies))
	for k, v := range o.headersCreateConversation {
		req.Header.Set(k, v)
	}
	resp, err := client.Do(req)
	if err != nil {
		return CreateConversationResponse{}, err
	}
	defer resp.Body.Close()
	bodyV, err := io.ReadAll(resp.Body)
	if err != nil {
		return CreateConversationResponse{}, err
	}
	if resp.StatusCode != 200 {
		return CreateConversationResponse{}, errors.New("Authentication failed: " + string(bodyV))
	}
	var response CreateConversationResponse
	err = json.Unmarshal(bodyV, &response)
	if err != nil {
		return CreateConversationResponse{}, err
	}
	if response.Result.Value == "UnauthorizedRequest" {
		return CreateConversationResponse{}, errors.New(response.Result.Message)
	}
	if value := resp.Header.Get("X-Sydney-Encryptedconversationsignature"); value != "" {
		response.SecAccessToken = value
	}
	if o.debug {
		log.Printf("%#v\n", response)
	}
	return response, nil
}

type Message struct {
	Data  string
	Error error
}
type ChatMessage struct {
	Arguments    []Argument `json:"arguments"`
	InvocationId string     `json:"invocationId"`
	Target       string     `json:"target"`
	Type         int        `json:"type"`
}
type Argument struct {
	OptionsSets           []string          `json:"optionsSets"`
	Source                string            `json:"source"`
	AllowedMessageTypes   []string          `json:"allowedMessageTypes"`
	SliceIds              []string          `json:"sliceIds"`
	Verbosity             string            `json:"verbosity"`
	Scenario              string            `json:"scenario"`
	TraceId               string            `json:"traceId"`
	RequestId             string            `json:"requestId"`
	IsStartOfSession      bool              `json:"isStartOfSession"`
	Message               ArgumentMessage   `json:"message"`
	Tone                  string            `json:"tone"`
	ConversationSignature any               `json:"conversationSignature"`
	Participant           Participant       `json:"participant"`
	SpokenTextMode        string            `json:"spokenTextMode"`
	ConversationId        string            `json:"conversationId"`
	PreviousMessages      []PreviousMessage `json:"previousMessages"`
}
type ArgumentMessage struct {
	Locale        string         `json:"locale"`
	Market        string         `json:"market"`
	Region        string         `json:"region"`
	LocationHints []LocationHint `json:"locationHints"`
	Author        string         `json:"author"`
	InputMethod   string         `json:"inputMethod"`
	Text          string         `json:"text"`
	MessageType   string         `json:"messageType"`
	RequestId     string         `json:"requestId"`
	MessageId     string         `json:"messageId"`
	ImageUrl      any            `json:"imageUrl"`
}
type Participant struct {
	Id string `json:"id"`
}
type PreviousMessage struct {
	Author      string `json:"author"`
	Description string `json:"description"`
	ContextType string `json:"contextType"`
	MessageType string `json:"messageType"`
	MessageId   string `json:"messageId"`
}

func (o *Sydney) AskStream(
	stopCtx context.Context,
	conversation CreateConversationResponse,
	prompt string,
	webpageContext string,
	imageURL string,
) <-chan Message {
	msgChan := make(chan Message)
	go func() {
		defer close(msgChan)
		client, err := MakeHTTPClient(o.proxy, 0)
		if err != nil {
			msgChan <- Message{
				Error: err,
			}
			return
		}
		messageID, err := uuid.NewUUID()
		if err != nil {
			msgChan <- Message{
				Error: err,
			}
			return
		}
		headers := CopyMap(o.headers)
		headers["Cookie"] = FormatCookieString(o.cookies)
		httpHeaders := map[string][]string{}
		for k, v := range headers {
			httpHeaders[k] = []string{v}
		}
		ctx, cancel := CreateTimeoutContext(10 * time.Second)
		defer cancel()
		connRaw, resp, err := websocket.Dial(ctx,
			o.wssURL+Ternary(conversation.SecAccessToken != "", "?sec_access_token="+
				url.QueryEscape(conversation.SecAccessToken), ""),
			&websocket.DialOptions{
				HTTPClient: client,
				HTTPHeader: httpHeaders,
			})
		if err != nil {
			msgChan <- Message{
				Error: err,
			}
			return
		}
		if resp.StatusCode != 101 {
			msgChan <- Message{
				Error: errors.New("cannot establish a websocket connection"),
			}
			return
		}
		defer connRaw.CloseNow()
		select {
		case <-stopCtx.Done():
			return
		default:
		}
		connRaw.SetReadLimit(-1)
		conn := &Conn{Conn: connRaw, debug: o.debug}
		err = conn.WriteWithTimeout([]byte(`{"protocol": "json", "version": 1}`))
		if err != nil {
			msgChan <- Message{
				Error: err,
			}
			return
		}
		conn.ReadWithTimeout()
		err = conn.WriteWithTimeout([]byte(`{"type": 6}`))
		if err != nil {
			msgChan <- Message{
				Error: err,
			}
			return
		}
		if o.noSearch {
			prompt += " #no_search"
		}
		chatMessage := ChatMessage{
			Arguments: []Argument{
				{
					OptionsSets:         o.optionsSetMap[o.conversationStyle],
					Source:              "cib",
					AllowedMessageTypes: o.allowedMessageTypes,
					SliceIds:            o.sliceIDs,
					Verbosity:           "verbose",
					Scenario:            "SERP",
					TraceId:             MustGenerateRandomHex(16),
					RequestId:           messageID.String(),
					IsStartOfSession:    true,
					Message: ArgumentMessage{
						Locale:        o.locale,
						Market:        o.locale,
						Region:        o.locale[len(o.locale)-2:],
						LocationHints: o.locationHints[o.locale],
						Author:        "user",
						InputMethod:   "Keyboard",
						Text:          prompt,
						MessageType:   []string{"Chat", "SearchQuery"}[RandIntInclusive(0, 1)],
						RequestId:     messageID.String(),
						MessageId:     messageID.String(),
						ImageUrl:      Ternary[any](imageURL == "", nil, imageURL),
					},
					Tone: o.conversationStyle,
					ConversationSignature: Ternary[any](conversation.ConversationSignature == "",
						nil, conversation.ConversationSignature),
					Participant:    Participant{Id: conversation.ClientId},
					SpokenTextMode: "None",
					ConversationId: conversation.ConversationId,
					PreviousMessages: []PreviousMessage{
						{
							Author:      "user",
							Description: webpageContext,
							ContextType: "WebPage",
							MessageType: "Context",
							MessageId:   "discover-web--page-ping-mriduna-----",
						},
					},
				},
			},
			InvocationId: "0",
			Target:       "chat",
			Type:         4,
		}
		chatMessageV, err := json.Marshal(&chatMessage)
		if err != nil {
			msgChan <- Message{
				Error: err,
			}
			return
		}
		err = conn.WriteWithTimeout(chatMessageV)
		if err != nil {
			msgChan <- Message{
				Error: err,
			}
			return
		}
		for {
			select {
			case <-stopCtx.Done():
				return
			default:
			}
			messages, err := conn.ReadWithTimeout()
			if err != nil {
				msgChan <- Message{
					Error: err,
				}
				return
			}
			if time.Now().Unix()%6 == 0 {
				err = conn.WriteWithTimeout([]byte(`{"type": 6}`))
				if err != nil {
					msgChan <- Message{
						Error: err,
					}
					return
				}
			}
			for _, msg := range messages {
				if msg == "" {
					continue
				}
				if !gjson.Valid(msg) {
					msgChan <- Message{
						Error: errors.New("malformed json"),
					}
					return
				}
				result := gjson.Parse(msg)
				if result.Get("type").Int() == 2 && result.Get("item.result.value").String() != "Success" {
					msgChan <- Message{
						Error: errors.New(result.Get("item.result.value").Raw + ": " +
							result.Get("item.result.message").Raw),
					}
					return
				}
				msgChan <- Message{
					Data: msg,
				}
				if result.Get("type").Int() == 2 {
					// finish the conversation
					return
				}
			}
		}
	}()
	return msgChan
}

type Conn struct {
	debug bool
	*websocket.Conn
}

func (o *Conn) WriteWithTimeout(v []byte) error {
	ctx, cancel := CreateTimeoutContext(5 * time.Second)
	defer cancel()
	bytes := append(v, []byte(string(delimiter))...)
	if o.debug {
		log.Println("sending: " + string(bytes))
	}
	return o.Write(ctx, websocket.MessageText, bytes)
}
func (o *Conn) ReadWithTimeout() ([]string, error) {
	ctx, cancel := CreateTimeoutContext(30 * time.Second)
	defer cancel()
	typ, v, err := o.Read(ctx)
	if err != nil {
		return nil, err
	}
	if typ != websocket.MessageText {
		return nil, nil
	}
	if len(v) == 0 {
		return nil, errors.New("no response from server")
	}
	str := string(v)
	arr := strings.Split(str, string(delimiter))
	if o.debug {
		for _, item := range arr {
			log.Println("receiving: " + item)
		}
	}
	return arr, nil
}
